### PR TITLE
add: group extensions

### PR DESCRIPTION
**Pull request by benkoshy**
### What is this?

Adds extension methods to the `Group` class - convenience methods.

### Why?

Simplifies the client code a tiny bit.  People are used to working with entities, or object ids. Mainly, they should not be too concerned about `ObjectIdCollections` nor using Linq to convert from one Enumerable type to another.

```c#
/// Before
IEnumerable<Entity> entities = polylines.Cast<Entity>();
IEnumerable<ObjectId> ids = entities.Select(e => e.ObjectId);
ObjectIdCollection idCollection = new ObjectIdCollection(ids.ToArray());
Group group = getGroup();
group.Append(idCollection);


// After:
IEnumerable<Entity> entities = polylines.Cast<Entity>();
group.append(entities);
group.append(ids);
```

### Problems with the PR:

* I was not able to properly integrate it in the project.
* I have not utilised it in my actual project. I have however looked over it carefully enough.
* Also no documentation has been added just yet.

### CODE REVIEW

File: src/Sources/Linq2Acad/Extensions/GroupExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    /// </summary>
+    public static  class GroupExtensions
+    {   /// <summary>
+        /// Appends ids to group. (Note, this crystalises the ids to an array internally).
+        /// </summary>
+        /// <param name="group"></param>
+        /// <param name="ids"></param>
+        public static void Append(this Group group, IEnumerable<ObjectId> ids)
+        {
+            ObjectIdCollection idCollection = new ObjectIdCollection(ids.ToArray());
```
Comment (wtertinek) [inline src/Sources/Linq2Acad/Extensions/GroupExtensions.cs:21]: ObjectIdCollections have to be disposed of. Best way to reliably call Dispose() is a using statement:

```cs
using (ObjectIdCollection idCollection = new ObjectIdCollection(ids.ToArray()))
{
  group.Append(idCollection);
}
```

Not disposing of ObjectIdCollections introduces a memory leak (and AutoCAD crashes unpredictably).

**Review by wtertinek — COMMENTED** [review]
The _GroupExtensions_ file has to be added to the csproj files. Best way to do this is to open the csproj files in a text editor and to add the following element:

```xml
<Compile Include="..\..\Sources\Linq2Acad\Extensions\GroupExtensions.cs">
  <Link>Extensions\GroupExtensions.cs</Link>
</Compile>
```

File: src/Sources/Linq2Acad/Extensions/GroupExtensions.cs (hunk 1 above)
Comment (wtertinek) [reply src/Sources/Linq2Acad/Extensions/GroupExtensions.cs:21]: Neither did I. Was one of the many surprises the AutoCAD API can give you. ;) I once did a larger batch process with some custom code and AutoCAD crashed unpredictably. Calling Dispose() on the ObjectIdCollection fixed the problem.

**Comment by benkoshy** [comment]
Hey Wolfgang,

I have:

- Added a c# code sample (and I snuck in a very hand extension method in order to test the code :D),. I have ran the code at it works for me.
- Updated all project references.


Issues with PR:
- VB.net code samples ignored.
- Latest methods not yet documented.
- Code sample does not exist in all VS projects - only one.

Would appreciate a review.

**Comment by benkoshy** [comment]
Further to this PR, would the following be a ridiculous proposal?

### Before

```c#
IEnumerable<Entity> entities = getEntities():
db.Groups.Create(groupName, entities)); 
```

### After
```c#
ObjectId[] ids= getIds();
db.Groups.Create(groupName, ids));   // no need to open DBObject and pass in entites.
```

**Comment by wtertinek** [comment]
> Further to this PR, would the following be a ridiculous proposal?
> ### Before
> 
> ```cs
> IEnumerable<Entity> entities = getEntities():
> db.Groups.Create(groupName, entities)); 
> ```
> 
> ### After
> 
> ```cs
> ObjectId[] ids= getIds();
> db.Groups.Create(groupName, ids));   // no need to open DBObject and pass in entites.
> ```

Good idea, please add this overload.

**Review by wtertinek — CHANGES_REQUESTED** [review]
The file src\AutoCAD_2022\Linq2Acad\GroupExtensions.cs should be removed, it is now in src\Sources\Linq2Acad\Extensions. The rest looks good :+1:.

**Comment by benkoshy** [comment]
> The file src\AutoCAD_2022\Linq2Acad\GroupExtensions.cs should be removed, it is now in src\Sources\Linq2Acad\Extensions. The rest looks good 👍.

Oops missed it during development.  Let me post hotfix and squash.

**Comment by benkoshy** [comment]
I think I have added all the changes. If you're happy after review, I"ll squash and rebase and push as a two separate commits: allowing you to cleanly merge.

Re: documentation, it seems like they are auto-generated? Meaning that I don't need to do anything more?

Many of these overloads could have applicability in other containers, but I am hesitant in making general PRs without a concrete use-case in a separate application.

**Comment by wtertinek** [comment]
Thank you Ben, very good work! Two separate commits to allow a clean merge is a good idea. 👍 Yes, documentation is auto-generated, I'll generate it after integrating the pull request. Also the generation of the Nuget packages is still a manual process. Please do the rebase, I'll finish the pull request afterwards.

**Review by wtertinek — APPROVED** [review]

**Comment by benkoshy** [comment]
> Thank you Ben, very good work! Two separate commits to allow a clean merge is a good idea. 👍 Yes, documentation is auto-generated, I'll generate it after integrating the pull request. Also the generation of the Nuget packages is still a manual process. Please do the rebase, I'll finish the pull request afterwards.

There will be a github action to manage nuget, the  complicating factor is the versions.  Surely there's a solution to that. But perhaps for anther day

**Comment by benkoshy** [comment]
The merge should be clean and the commits atomic.

**Problems with PR:**

- I have not included a Visual Basic code sample.

**Review by wtertinek — COMMENTED** [review]
Somehow the using statements for the ObjectIdColletions got lost.

**Comment by benkoshy** [comment]
> Somehow the using statements for the ObjectIdColletions got lost.

I"ll post a hot fix. not sure how that happened.

**Comment by wtertinek** [comment]
> **Update:** force pushed and squashed the fix. Also I have applied a code sample only for 2022. I suppose it should be applied to all versions. I shall do that after the merge or before?

(Sorry for my late response) Yes, please add the sample code as well (before the merge). If you add a new file, please add it the same way you added the extension method: add the file under src/Sources and manually edit the csproj files to add the file to the project.

Unfortunately adding files is very cumbersome. Improving the project structure (redrucing the number of solutions) is definitely the next thing to do (I found [this](https://stackoverflow.com/a/61991596) on SO).

**Comment by benkoshy** [comment]
> > **Update:** force pushed and squashed the fix. Also I have applied a code sample only for 2022. I suppose it should be applied to all versions. I shall do that after the merge or before?
> 
> (Sorry for my late response) Yes, please add the sample code as well (before the merge). If you add a new file, please add it the same way you added the extension method: add the file under src/Sources and manually edit the csproj files to add the file to the project.

I can confirm that the sample code changes have been applied to all projects without me doing anything. This was because the `SampleCode.cs` had already been added to the project. All I did was change the source directly.

Hence it should be ready to merge.

If you find any issues, do not hesitate to send it back.


> 
> Unfortunately adding files is very cumbersome. Improving the project structure (redrucing the number of solutions) is definitely the next thing to do (I found [this](https://stackoverflow.com/a/61991596) on SO).

**Comment by wtertinek** [comment]
Thanks Ben, looks good. I'll upate the documentation and create the Nuget packages in the next few days.

**Pull request merged**